### PR TITLE
feat: move insights between dashboards when many-to-many

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

Currently when you move an insight between dashboards we duplicate the insight onto the new dashboard and soft-delete the original dashboard

There is no reason to create a new insight once insights can live on multiple dashboards

## Changes

Adds a new moveToDashboard method behind a flag which updates the insight with the new set of dashboards

### with the flag off

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

moving insights between dashboards with and without the flag on

### CODE REVIEW

File: frontend/src/scenes/insights/insightLogic.ts [hunk 1]
```diff
@@ -356,6 +356,12 @@ export const insightLogic = kea<insightLogicType>({
                 ...insight,
             }),
             setInsightMetadata: (state, { metadata }) => ({ ...state, ...metadata }),
+            [dashboardsModel.actionTypes.updateDashboardItem]: (state, { item }) => {
+                if (item.short_id === state.short_id) {
+                    return { ...state, dashboards: item.dashboards }
+                }
+                return state
+            },
```
Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:364]: I wonder if we shouldn't just update everything? Any chance we need to? What could go wrong if we do?

Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:364]: I wonder if we shouldn't just update everything? Any chance we need to? What could go wrong if we do?

**Review by mariusandra — CHANGES_REQUESTED** [review]
The food came on the train, so submitting the review. I'd improve the typing and some wiring a bit.

File: frontend/src/lib/components/InsightCard/InsightCard.tsx [hunk 2]
```diff
@@ -139,7 +140,8 @@ export interface InsightCardProps extends React.HTMLAttributes<HTMLDivElement> {
     refresh?: () => void
     rename: () => void
     duplicate: () => void
-    moveToDashboard?: (dashboardId: DashboardType['id']) => void
+    moveToDashboardOld?: (dashboardId: DashboardType['id']) => void
+    moveToDashboard?: (dashboardIdentifiers: Pick<DashboardType, 'id' | 'name'>) => void
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply frontend/src/lib/components/InsightCard/InsightCard.tsx:144]: 🤔 this feels more expressive than a new type to me... this method takes specifically the id and name of a dashboard... I guess the alternative is to hide that from consumers and pass a dashboard in

File: frontend/src/scenes/dashboard/dashboardLogic.test.ts [hunk 3]
```diff
@@ -5,48 +5,65 @@ import _dashboardJson from './__mocks__/dashboard.json'
...
 import { resumeKeaLoadersErrors, silenceKeaLoadersErrors } from '~/initKea'
 import { useMocks } from '~/mocks/jest'
 import { dayjs, now } from 'lib/dayjs'
 import posthog from 'posthog-js'
+import { teamLogic } from 'scenes/teamLogic'
+import { featureFlagLogic } from 'lib/logic/featureFlagLogic'
 
 const dashboardJson = _dashboardJson as any as DashboardType
 
+const dashboardResult = (dashboardsRelation: number[]): DashboardType => {
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply frontend/src/scenes/dashboard/dashboardLogic.test.ts:18]: Hmm, probably exposing laziness in the tests...

File: frontend/src/scenes/insights/insightLogic.ts (hunk 1 above)
Comment (pauldambra) [reply frontend/src/scenes/insights/insightLogic.ts:364]: My understanding was that this is being called for every `updateDashboardItem` not just those for this insight/dashboard combo. I'll confirm that

Comment (pauldambra) [reply frontend/src/scenes/insights/insightLogic.ts:364]: Ah...

You mean

```
    [dashboardsModel.actionTypes.updateDashboardItem]: (state, { item }) => {
        if (item.short_id === state.short_id) {
            return { ...item }
        }
        return state
    },
```

???

yep... reasonable

File: frontend/src/scenes/dashboard/dashboardLogic.ts [hunk 4]
```diff
@@ -95,6 +95,7 @@ export const dashboardLogic = kea<dashboardLogicType<DashboardLogicProps>>({
     },
 
     loaders: ({ actions, props }) => ({
+        // TODO this is a terrible name... it is "dashboard" but there's a "dashboard" reducer ¯\_(ツ)_/¯
```
Comment (mariusandra) [inline frontend/src/scenes/dashboard/dashboardLogic.ts:98]: 🙈

File: frontend/src/scenes/insights/insightLogic.ts [hunk 5]
```diff
@@ -356,6 +356,12 @@ export const insightLogic = kea<insightLogicType>({
                 ...insight,
             }),
             setInsightMetadata: (state, { metadata }) => ({ ...state, ...metadata }),
+            [dashboardsModel.actionTypes.updateDashboardItem]: (state, { item }) => {
+                if (item.short_id === state.short_id) {
+                    return { ...item }
```
Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:361]:
```suggestion
                    return item
```

**Review by mariusandra — APPROVED** [review]

**Pull request merged**